examples/pennylane_run_qaoa_maxcut.py

```diff
@@ -0,0 +1,288 @@
+#
+# .. math::
+#   U_{B_i} &= e^{-i\beta_iB} = \prod_{j=1}^n e^{-i\beta_i\sigma_x^j} \\
+#   U_{C_i} &= e^{-i\gamma_iC} = \prod_{\text{edge (j,k)}} e^{-i\gamma_i(1-\sigma_z^j\sigma_z^k)/2}
+#
+# These can be implemented on a quantum circuit using the gates depicted below (up to an irrelevant constant
+# that gets absorbed into the parameters).
+#
+#
+# .. figure:: ../../examples/figures/qaoa_operators.png
+#    :align: center
+#    :scale: 100%
+#    :alt: qaoa_operators
+#
+#
+# Let :math:`\langle \boldsymbol{\gamma},
+# \boldsymbol{\beta} | C | \boldsymbol{\gamma},\boldsymbol{\beta} \rangle` be the expectation of our objective.
+# In the next section, we will use PennyLane to sample this expectation value and perform classical optimization
+# over the parameters. This will specify a state :math:`|\boldsymbol{\gamma},\boldsymbol{\beta}\rangle` which is
+# likely to yield an approximately optimal partition :math:`|z\rangle` upon performing a measurement in the
+# computational basis.
+# In the case of the graph shown above, we want to measure either 0101 or 1010 from our state.
+#
+# .. figure:: ../../examples/figures/qaoa_optimal_state.png
+#   :align: center
+#   :scale: 60%
+#   :alt: optimal_state
+#
+#
+# 2 Implementing QAOA in PennyLane
+# ----------------------------------
+#
+# 2.2 Imports and setup
+# ~~~~~~~~~~~~~~~~~~~~~
+#
+# To get started, we import PennyLane along with the PennyLane-provided
+# version of NumPy.
+
+
+import pennylane as qml
+from pennylane import numpy as np
+
+
+##############################################################################
+# 2.3 Operators
+# ~~~~~~~~~~~~~
+# We specify the number of qubits (vertices) with ``n_wires`` and
+# compose the unitary operators using the definitions
+# above. :math:`U_B` operators act on each of the wires, while :math:`U_C`
+# operators act on wires whose corresponding vertices are joined by an edge in
+# the graph. We also define the graph using
+# the list ``graph``, which contains the tuples of vertices defining
+# each edge in the graph.
+
+n_wires = 4
+graph = [(0, 1), (0, 3), (1, 2), (2, 3)]
+
+# unitary operator U_B with parameter beta
+def U_B(beta):
+    for wire in range(n_wires):
+        qml.RX(2 * beta, wires=wire)
+
+
+# unitary operator U_C with parameter gamma
+def U_C(gamma):
+    for edge in graph:
+        wire1 = edge[0]
+        wire2 = edge[1]
+        qml.CNOT(wires=[wire1, wire2])
+        qml.RZ(gamma, wires=wire2)
+        qml.CNOT(wires=[wire1, wire2])
+
+
+##############################################################################
+# With some foresight, we notice that we will need a way to sample
+# a measurement of multiple qubits in the computational basis, so we define
+# a Hermitian operator to do this. The eigenvalues of the operator are
+# the joint qubit measurement values in integer form.
+
+
+def comp_basis_measurement(wires):
+    n_wires = len(wires)
+    return qml.Hermitian(np.diag(range(2 ** n_wires)), wires=wires)
```

Comment:
Possibility that default qubit plugin doesn't permit this approach: investigating further

Reply:
Looks like PennyLane does not correctly take into account qubit correlations on statevector simulators (but this works fine for hardware/hardware simulators, since they explicitly return the samples). We might have to re-architect this.

Reply:
^ This seems an important issue to address: algorithms often require more than just expectation values and variances!